Command to migrate DB data from `v1.0.0` to `v2.0.0`

josecelano opened the pull request:
Depends on: https://github.com/torrust/torrust-index-backend/pull/87

I've created the basic scaffolding of the command. My plan is:

- Migrate all the table records.
- Insert torrents in DB from `uploads` dir.

Potential problems are:

- AUTOINCREMENTS
- Foreign keys

**UPDATE 2022-11-30**

### Tasks

- [x] Transfer `torrent_categories`
- [x] Transfer `torrent_user`
- [x] Allow login for the transferred users using the Pbkdf2 for hashing the password
- [x] Use date and time in `today_iso8601` function instead of only the date
- [x] Transfer `torrent_tracker_keys`
- [x] Remove unused content from dir `upgrades`. We do not need the MySQL stuff for this migration. There was no MySQL support for version `v1.0.0`
- [x] Transfer `torrent_torrents`. See below the subtasks for this task
- [x] Transfer `torrent_torrent_files`. It was not used in `v1.0.0`
- [x] Make `registration_date` optional and leave it empty for imported users
- [x] Add a new DB field `imported_date` to the `torrust_users` table with the importation date for user records imported with this command
- [x] Refactor
- [x] Add automated tests
- [x] Intensive testing
- [x] Write documentation explaining how to upgrade
- [x] Rename `destiny` DB to `target` DB.

### Subtasks from issue #84 

This rework is needed because @ldpr found a bug in the current version that affects the upgrader. We have to fix that bug and update the "upgrader".
- [x] Fix issue #84 
- [x] Rework: remove the `torrust_torrent_tracker_stats::tracker_url` column. **DISCARDED**.
- [x] Populate table `torrust_torrent_tracker_stats` with one record per torrent.

### Transfer torrents subtasks

- [x] Fill table `torrust_torrents`
- [x] Fill table `torrust_torrent_files`
- [x] Fill table `torrust_torrent_announce_urls`
- [x] Fill table `torrust_torrent_info`

### Testing subtasks

Assertions for integration test:

- [x] `torrust_users` table
- [x] `torrust_user_authentication` table
- [x] `torrust_user_profiles` table
- [x] `torrust_tracker_keys` table
- [x] `torrust_torrents` table
- [x] `torrust_torrent_files` table
- [x] `torrust_torrent_info` table
- [x] `torrust_torrent_announce_urls` table

Some cases to test:

- [x] Torrent with `pieces`.

Using a single tracker URL or multiple URLs:

- [x] Torrent with `announce` (single tracker url).
- [x] Torrent with `announce_list` (multiple tracker URLs).

Containing one file or multiple files:

- [x] Torrent with one file.
- [x] Torrent with multiple files.

Add tests for the application behaviour changed:

- [x] User can use have the password hashed with "argon2id" or a "pbkdf2-sha256". Add a test for the "verify_password" function.

Things I'm not going to test:

- [ ] Torrent with MD5 checksum?. I do not know how to include this MD5 checksum in the torrent file.
- [ ] Torrent with `root_hash` ([BEP-0030](https://www.bittorrent.org/beps/bep_0030.html)).
- [ ] The user registration date can be null for users that have been imported. We do not test it for MySQL since the upgrader is only for SQLite.

josecelano commented:
hi @WarmBeer, I'm working on transferring user data. I've done all except the password. Since the hashing algorithm has changed from Pbkdf2 to Argon2, I have to think about how to move the data. 

I think the best way to do it would be to add a new column to the database with the type of hashing used and use the old method until the user changes the password.

What do you think?

cc @da2ce7

WarmBeer commented:
Hey @josecelano ,

I'm think we should revert back to Pbkdf2, since we are dropping username/password authentication in the future anyway.

da2ce7 commented:
> hi @WarmBeer, I'm working on transferring user data. I've done all except the password. Since the hashing algorithm has changed from Pbkdf2 to Argon2, I have to think about how to move the data.
> 
> I think the best way to do it would be to add a new column to the database with the type of hashing used and use the old method until the user changes the password.
> 
> What do you think?
> 
> cc @da2ce7

Users do not need to change their password to change the password hashing algorithm: The only need to present with a valid password.

1. Save imported user hashed-passwords and nonces into a new "Pbkdf2" table.
2. User tries to login: Find the correct table: "Pbkdf2" or "Argon2". Confirm password normally.

Optional: Generate new password nonce and calculated password hash in "Argon2". Make an atomic transaction that inserts the new hashed-password-and-nounce, and drops the old one.

I think that it is too-late to move away from "Argon2", as there might be people who use our unreleased v.2 that have active users in this format.

josecelano commented:
> > hi @WarmBeer, I'm working on transferring user data. I've done all except the password. Since the hashing algorithm has changed from Pbkdf2 to Argon2, I have to think about how to move the data.
> > I think the best way to do it would be to add a new column to the database with the type of hashing used and use the old method until the user changes the password.
> > What do you think?
> > cc @da2ce7
> 
> Users do not need to change their password to change the password hashing algorithm: The only need to present with a valid password.
> 
> 1. Save imported user hashed-passwords and nonces into a new "Pbkdf2" table.
> 2. User tries to login: Find the correct table: "Pbkdf2" or "Argon2". Confirm password normally.
> 
> Optional: Generate new password nonce and calculated password hash in "Argon2". Make an atomic transaction that inserts the new hashed-password-and-nounce, and drops the old one.
> 
> I think that it is too-late to move away from "Argon2", as there might be people who use our unreleased v.2 that have active users in this format.

hi @WarmBeer @da2ce7 ,

I did not know we were using the [PHC string format](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md) since version v1.0.0. That means we have values like this in the password hash value in the DB.

```text
$pbkdf2-sha256$i=10000,l=32$DGEHp7+y+kvTBUt3XdqFgA$JsK3jdz4Yz4hUPmWmPCyWk+NlgkdN9TCkbcjeQV68fA
```

I only had to import the hash from the old DB to the new DB field and change the `login` controller to use a different verification depending on the prefix of the hash value.

NOTE: that means we do not need to have two columns for both hash types and a third one with the name of the algorithm used.

In the previous example: `pbkdf2-sha256`.
That's the hash algorithm ID.
And this is the password verification in the login:

```rust
pub fn verify_password(
    password: &[u8],
    user_authentication: &UserAuthentication,
) -> Result<(), ServiceError> {
    // wrap string of the hashed password into a PasswordHash struct for verification
    let parsed_hash = PasswordHash::new(&user_authentication.password_hash)?;

    match parsed_hash.algorithm.as_str() {
        "argon2id" => {
            if Argon2::default()
                .verify_password(password, &parsed_hash)
                .is_err()
            {
                return Err(ServiceError::WrongPasswordOrUsername);
            }

            Ok(())
        }
        "pbkdf2-sha256" => {
            if Pbkdf2.verify_password(password, &parsed_hash).is_err() {
                return Err(ServiceError::WrongPasswordOrUsername);
            }

            Ok(())
        }
        _ => Err(ServiceError::WrongPasswordOrUsername),
    }
}
```

We can upgrade the hash when the user logs in if we want. Should I do it now in this PR? Are we going to do it? I would do it to keep things simple. But I would do it in a new minor release (v2.1.0). It's an optional feature, and I would prefer to make the minimum amount of changes in this upgrade. We are already doing a lot of changes.

da2ce7 commented:
@josecelano  Assisting the current date for the registration data is a loss of information. I suggest that we have that field to be optional, and have a new entry that contains the date of import.

josecelano commented:
> @josecelano Assisting the current date for the registration data is a loss of information. I suggest that we have that field to be optional, and have a new entry that contains the date of import.

hi @da2ce7, Yes, It does not make sense at all. Probably I was thinking about creating a new site as if the users were registered at that moment (but the migration has to be transparent to the user). I realised later, but I have not changed it yet. I think your suggestion is the right way to do it.

Besides, I wanted to propose another change for the future. It would be nice to have "created_at" and "updated_at" fields in all the tables. I think that could help to debug bugs. Ideally, we could get that info from the log, but it's very useful information when you are trying to know what happened, and it's very handy to have it on the table.

josecelano commented:
hi @WarmBeer It seem this table was not used at all:

```
CREATE TABLE IF NOT EXISTS torrust_torrent_files (
    file_id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    torrent_id INTEGER NOT NULL,
    number INTEGER NOT NULL,
    path VARCHAR(255) NOT NULL,
    length INTEGER NOT NULL,
    FOREIGN KEY(torrent_id) REFERENCES torrust_torrents(torrent_id)
)
```

I see the migration in this commit 6d66a0ad8cd81c71ec2029b43cd59737936d265c, but I do not see any DB query to use it. Could you confirm it?

WarmBeer commented:
Hi @josecelano ,

The table is being used here: https://github.com/torrust/torrust-index-backend/blob/e8a6fe75654e769d4d12b49d77bae1ff159bc13c/src/databases/mysql.rs#L416

josecelano commented:
> Hi @josecelano ,
> 
> The table is being used here:
> 
> https://github.com/torrust/torrust-index-backend/blob/e8a6fe75654e769d4d12b49d77bae1ff159bc13c/src/databases/mysql.rs#L416
> 
> Which is part of this function:
> 
> https://github.com/torrust/torrust-index-backend/blob/e8a6fe75654e769d4d12b49d77bae1ff159bc13c/src/databases/mysql.rs#L365

Sorry @WarmBeer, I did not explain it well. I mean, it was created before release `v1.0.0`, but it was not used after that release. Since I'm only working on migrating from v1.0.0 to v2.0.0, that means that table should be empty. I'm using the index with the backend v1.0.0 and I can upload a torrent but that table remains empty.

WarmBeer commented:
Oh sorry @josecelano, you are right. This table is not used at all prior to V2.

josecelano commented:
hi @WarmBeer, two questions:

1. Were all torrents in v1.0.0 always public? I do not see any column to mark them as private.
2. Inhohases were lowercase in v1.0.0 and uppercase in v2.0.0. I suppose it does not matter. I've concerted them to uppercase.

josecelano commented:
> hi @WarmBeer, two questions:
> 
> 1. Were all torrents in v1.0.0 always public? I do not see any column to mark them as private.
> 2. Inhohases were lowercase in v1.0.0 and uppercase in v2.0.0. I suppose it does not matter. I've converted them to uppercase.

hi @WarmBeer @da2ce7, I'm done with transferring the data. Now I want to refactor a little bit and make some tests with more data and cases. And improve the documentation.

@WarmBeer a part for the previous questions (1 and 2) I have one more:

3. It seems that if the torrent has "announce" and "announce_list" (BEP-0012 for multiple trackers), we only add the "announce" (tracker URL) to the `torrust_torrent_announce_urls` table. But the [BEP-012](https://www.bittorrent.org/beps/bep_0012.html) says "announce_list" should have higher priority.

![image](https://user-images.githubusercontent.com/58816/200806093-71f4910e-d8d8-4f90-85b3-25ab81bcb20e.png)

[Our code for SQlite](https://github.com/torrust/torrust-index-backend/blob/develop/src/databases/sqlite.rs#L445-L468).

WarmBeer commented:
Hi @josecelano , Sorry for the late response.

>  1. Were all torrents in v1.0.0 always public? I do not see any column to mark them as private.

In both V1 and V2, torrents are always public on the torrent index. There is currently nothing in place to restrict viewing rights. 

But If you are talking about the .torrent metadata setting, then no. The private flag would just be inside the .torrent file for V1, where in V2 it is in the database.

> 2. Inhohases were lowercase in v1.0.0 and uppercase in v2.0.0. I suppose it does not matter. I've concerted them to uppercase.

I do not actively remember making this change and what drove that decision. But I suppose it doesn't matter 😅.

> 3. It seems that if the torrent has "announce" and "announce_list" (BEP-0012 for multiple trackers), we only add the "announce" (tracker URL) to the torrust_torrent_announce_urls table. But the [BEP-012](https://www.bittorrent.org/beps/bep_0012.html) says "announce_list" should have higher priority.

Good find. This is indeed a mistake. I will open a PR to prioritize the announce_list field over the announce field.

josecelano commented:
> Hi @josecelano , Sorry for the late response.
> 
> > 1. Were all torrents in v1.0.0 always public? I do not see any column to mark them as private.
> 
> In both V1 and V2, torrents are always public on the torrent index. There is currently nothing in place to restrict viewing rights.
> 
> But If you are talking about the .torrent metadata setting, then no. The private flag would just be inside the .torrent file for V1, where in V2 it is in the database.
> 
> > 2. Inhohases were lowercase in v1.0.0 and uppercase in v2.0.0. I suppose it does not matter. I've concerted them to uppercase.
> 
> I do not actively remember making this change and what drove that decision. But I suppose it doesn't matter sweat_smile.
> 
> > 3. It seems that if the torrent has "announce" and "announce_list" (BEP-0012 for multiple trackers), we only add the "announce" (tracker URL) to the torrust_torrent_announce_urls table. But the [BEP-012](https://www.bittorrent.org/beps/bep_0012.html) says "announce_list" should have higher priority.
> 
> Good find. This is indeed a mistake. I will open a PR to prioritize the announce_list field over the announce field.

1. OK, thank you. I've fixed it [here](https://github.com/torrust/torrust-index-backend/pull/77/commits/694bbafbb414ab9ff433b563bbb4cb4d86a87eca).
2. It does matter, but we should always use the same in our DB and UI. I prefer the lowercase version, but let me know which one you prefer to change the upgrader.
3. OK, I will fix it in the upgrader.

WarmBeer commented:
> > Hi @josecelano , Sorry for the late response.
> > > 1. Were all torrents in v1.0.0 always public? I do not see any column to mark them as private.
> > 
> > 
> > In both V1 and V2, torrents are always public on the torrent index. There is currently nothing in place to restrict viewing rights.
> > But If you are talking about the .torrent metadata setting, then no. The private flag would just be inside the .torrent file for V1, where in V2 it is in the database.
> > > 2. Inhohases were lowercase in v1.0.0 and uppercase in v2.0.0. I suppose it does not matter. I've concerted them to uppercase.
> > 
> > 
> > I do not actively remember making this change and what drove that decision. But I suppose it doesn't matter sweat_smile.
> > > 3. It seems that if the torrent has "announce" and "announce_list" (BEP-0012 for multiple trackers), we only add the "announce" (tracker URL) to the torrust_torrent_announce_urls table. But the [BEP-012](https://www.bittorrent.org/beps/bep_0012.html) says "announce_list" should have higher priority.
> > 
> > 
> > Good find. This is indeed a mistake. I will open a PR to prioritize the announce_list field over the announce field.
> 
> 1. OK, thank you. I've fixed it [here](https://github.com/torrust/torrust-index-backend/pull/77/commits/694bbafbb414ab9ff433b563bbb4cb4d86a87eca).
> 2. It does matter, but we should always use the same in our DB and UI. I prefer the lowercase version, but let me know which one you prefer to change the upgrader.
> 3. OK, I will fix it in the upgrader.

I usually prefer lowercase in the database and uppercase in the frontend.

WarmBeer commented:
https://github.com/torrust/torrust-index-backend/pull/80

josecelano commented:
hi @WarmBeer, I've added a test to check that one torrent is transferred well from the old DB and file to the new. The torrent I use in the test is a torrent:

- With one file.
- Using "pieces" instead of "root_hash".
- Using "announce_list" instead of "announce".
- No MD5 checksum.

I want to add a second test with a new torrent:

- With multiple files (folder).
- Using "root_hash" instead of "pieces".
- Using "announce" instead of "announce_list".
- With MD5 checksum.

But I do not know how to force those options (root_hash, announce, and MD5 checksum) using the BitTorrent client:

![image](https://user-images.githubusercontent.com/58816/201323154-3ed64088-878d-4cac-9a41-43930cf75df8.png)

I want to cover all cases with two torrents if possible. Do you know how I can create a torrent with those fields?

That's the last automatic test I want to add. After that, I will test the migrated DB with a running app.

WarmBeer commented:
> hi @WarmBeer, I've added a test to check that one torrent is transferred well from the old DB and file to the new. The torrent I use in the test is a torrent:
> 
> * With one file.
> * Using "pieces" instead of "root_hash".
> * Using "announce_list" instead of "announce".
> * No MD5 checksum.
> 
> I want to add a second test with a new torrent:
> 
> * With multiple files (folder).
> * Using "root_hash" instead of "pieces".
> * Using "announce" instead of "announce_list".
> * With MD5 checksum.
> 
> But I do not know how to force those options (root_hash, announce, and MD5 checksum) using the BitTorrent client:
> 
> ![image](https://user-images.githubusercontent.com/58816/201323154-3ed64088-878d-4cac-9a41-43930cf75df8.png)
> 
> I want to cover all cases with two torrents if possible. Do you know how I can create a torrent with those fields?
> 
> That's the last automatic test I want to add. After that, I will test the migrated DB with a running app.

Hey @josecelano ,

I'm sorry but I don't know of any alternative Torrent file creation tools that do support these extensions or Bittorrent V2. Even qBittorrent does not show any options for creating V2 torrent files.

josecelano commented:
hey @WarmBeer, I did not know we don't support [The BitTorrent Protocol Specification v2 - BEP52](https://www.bittorrent.org/beps/bep_0052.html).

I'm going to assume we only have torrents in version V1. If that's the case I'm almost doe with tests.

On the other hand, do you know why the `md5sum` field is always empty in `TorrentInfo`?:

```rust
pub struct TorrentInfo {
    pub name: String,
    #[serde(default)]
    pub pieces: Option<ByteBuf>,
    #[serde(rename = "piece length")]
    pub piece_length: i64,
    #[serde(default)]
    pub md5sum: Option<String>,
    #[serde(default)]
    pub length: Option<i64>,
    #[serde(default)]
    pub files: Option<Vec<TorrentFile>>,
    #[serde(default)]
    pub private: Option<u8>,
    #[serde(default)]
    pub path: Option<Vec<String>>,
    #[serde(default)]
    #[serde(rename = "root hash")]
    pub root_hash: Option<String>,
}
```

I do not see any reference to that field in https://www.bittorrent.org/.

josecelano commented:
Hi @WarmBeer @da2ce7, I think the PR is ready to review. I still want to run some more manual tests, like running the application with the migrated DB. But I'm not planning more reactors or changes.

I was planning to add colours to the output, but it is not worth it because it's a one-use script.

Regarding automated testing, I've added torrents that cover all the cases I wanted to test. Maybe I could have added one torrent per test, but we can easily add more torrents if you think it's better. I'm not completely happy with the way I'm testing the command. I would like to have one test in the output for each assert (high-level assert, for example, the torrent in the destiny DB). I can refactor the test to use a new clean state for each test bu I decided to setup the state only once. The reason is I wanted to create one single output tests, considering the input the dump of the source DB and the torrent files and the output the dump of the destiny DB. But in the end, I thought using a SQL file for the inputs and outputs was not flexible enough in this case.

Let me know if you think the documentation is enough.

This is a sample output when you run the script:

```s
Upgrading data from version v1.0.0 to v2.0.0 ...
Running migrations in destiny database...
Truncating all tables in destiny database ...
Transferring categories ...
[v1] categories: [CategoryRecord { category_id: 1, name: "movies" }, CategoryRecord { category_id: 2, name: "tv shows" }, CategoryRecord { category_id: 3, name: "games" }, CategoryRecord { category_id: 4, name: "music" }, CategoryRecord { category_id: 5, name: "software" }, CategoryRecord { category_id: 6, name: "another" }]
[v2] reset categories sequence result SqliteQueryResult { changes: 1, last_insert_rowid: 0 }
[v2] adding category "movies" with id 1 ...
[v2] category: 1 "movies" added.
[v2] adding category "tv shows" with id 2 ...
[v2] category: 2 "tv shows" added.
[v2] adding category "games" with id 3 ...
[v2] category: 3 "games" added.
[v2] adding category "music" with id 4 ...
[v2] category: 4 "music" added.
[v2] adding category "software" with id 5 ...
[v2] category: 5 "software" added.
[v2] adding category "another" with id 6 ...
[v2] category: 6 "another" added.
[v2] categories: [CategoryRecordV2 { category_id: 6, name: "another" }, CategoryRecordV2 { category_id: 3, name: "games" }, CategoryRecordV2 { category_id: 1, name: "movies" }, CategoryRecordV2 { category_id: 4, name: "music" }, CategoryRecordV2 { category_id: 5, name: "software" }, CategoryRecordV2 { category_id: 2, name: "tv shows" }]
Transferring users ...
[v2][torrust_users] adding user with username "josecelano" and id 1 ...
[v2][torrust_users] user: 1 "josecelano" added.
[v2][torrust_user_profiles] adding user profile for user with username "josecelano" and id 1 ...
[v2][torrust_user_profiles] user profile added for user with username "josecelano" and id 1.
[v2][torrust_user_authentication] adding password hash ("$pbkdf2-sha256$i=10000,l=32$DGEHp7+y+kvTBUt3XdqFgA$JsK3jdz4Yz4hUPmWmPCyWk+NlgkdN9TCkbcjeQV68fA") for user id (1) ...
[v2][torrust_user_authentication] password hash ("$pbkdf2-sha256$i=10000,l=32$DGEHp7+y+kvTBUt3XdqFgA$JsK3jdz4Yz4hUPmWmPCyWk+NlgkdN9TCkbcjeQV68fA") added for user id (1).
[v2][torrust_users] adding user with username "josecelano2" and id 2 ...
[v2][torrust_users] user: 2 "josecelano2" added.
[v2][torrust_user_profiles] adding user profile for user with username "josecelano2" and id 2 ...
[v2][torrust_user_profiles] user profile added for user with username "josecelano2" and id 2.
[v2][torrust_user_authentication] adding password hash ("$pbkdf2-sha256$i=10000,l=32$qsCNNyFroO2rEi673efsaQ$DHMQp2VTm0KcGzSlOML75hKzKIzlYrjhaM3Xlss6Yjc") for user id (2) ...
[v2][torrust_user_authentication] password hash ("$pbkdf2-sha256$i=10000,l=32$qsCNNyFroO2rEi673efsaQ$DHMQp2VTm0KcGzSlOML75hKzKIzlYrjhaM3Xlss6Yjc") added for user id (2).
[v2][torrust_users] adding user with username "josecelano3" and id 3 ...
[v2][torrust_users] user: 3 "josecelano3" added.
[v2][torrust_user_profiles] adding user profile for user with username "josecelano3" and id 3 ...
[v2][torrust_user_profiles] user profile added for user with username "josecelano3" and id 3.
[v2][torrust_user_authentication] adding password hash ("$pbkdf2-sha256$i=10000,l=32$efa28HFc5zDUqdg3XVEtRg$wE6pRxW3GhVKJ357R1AyLHyDUSkiIzUpLU3SG3N9poU") for user id (3) ...
[v2][torrust_user_authentication] password hash ("$pbkdf2-sha256$i=10000,l=32$efa28HFc5zDUqdg3XVEtRg$wE6pRxW3GhVKJ357R1AyLHyDUSkiIzUpLU3SG3N9poU") added for user id (3).
Transferring tracker keys ...
[v2][torrust_users] adding the tracker key with id 1 ...
[v2][torrust_tracker_keys] tracker key with id 1 added.
[v2][torrust_users] adding the tracker key with id 2 ...
[v2][torrust_tracker_keys] tracker key with id 2 added.
Transferring torrents ...
[v2][torrust_torrents] adding the torrent: 17 ...
[v2][torrust_torrents] torrent with id 17 added.
[v2][torrust_torrent_files] adding torrent files
[v2][torrust_torrent_files][single-file-torrent] adding torrent file "Bonnard,_Pierre_-_Early_Spring._Little_Fauns.jpg" with length Some(9219566) ...
[v2][torrust_torrent_files][single-file-torrent] torrent file insert result: Ok(502)
[v2][torrust_torrent_info] adding the torrent info for torrent id 17 ...
[v2][torrust_torrents] torrent info insert result: Ok(17).
[v2][torrust_torrent_announce_urls] adding the torrent announce url for torrent id 17 ...
[v2][torrust_torrent_announce_urls][announce] adding the torrent announce url for torrent id 17 ...
[v2][torrust_torrent_announce_urls][announce] torrent announce url insert result Ok(173) ...
[v2][torrust_torrents] adding the torrent: 18 ...
[v2][torrust_torrents] torrent with id 18 added.
[v2][torrust_torrent_files] adding torrent files
[v2][torrust_torrent_files][single-file-torrent] adding torrent file "Boilly,_Louis-Leopold_-_A_Game_of_Billiards.jpg" with length Some(16425317) ...
[v2][torrust_torrent_files][single-file-torrent] torrent file insert result: Ok(503)
[v2][torrust_torrent_info] adding the torrent info for torrent id 18 ...
[v2][torrust_torrents] torrent info insert result: Ok(18).
[v2][torrust_torrent_announce_urls] adding the torrent announce url for torrent id 18 ...
[v2][torrust_torrent_announce_urls][announce] adding the torrent announce url for torrent id 18 ...
[v2][torrust_torrent_announce_urls][announce] torrent announce url insert result Ok(174) ...
[v2][torrust_torrents] adding the torrent: 19 ...
[v2][torrust_torrents] torrent with id 19 added.
[v2][torrust_torrent_files] adding torrent files
[v2][torrust_torrent_files][single-file-torrent] adding torrent file "Bellotto,_Bernardo_-_View_of_Pirna_from_the_Sonnenstein_Castle.jpg" with length Some(37783545) ...
[v2][torrust_torrent_files][single-file-torrent] torrent file insert result: Ok(504)
[v2][torrust_torrent_info] adding the torrent info for torrent id 19 ...
[v2][torrust_torrents] torrent info insert result: Ok(19).
[v2][torrust_torrent_announce_urls] adding the torrent announce url for torrent id 19 ...
[v2][torrust_torrent_announce_urls][announce] adding the torrent announce url for torrent id 19 ...
[v2][torrust_torrent_announce_urls][announce] torrent announce url insert result Ok(175) ...
Torrents transferred
```

When you run the test, you have a copy of both the source and destiny databases in the dir:

```s
└── upgrades
    ├── from_v1_0_0_to_v2_0_0
    │   ├── fixtures
    │   │   ├── database
    │   │   │   └── v1.0.0
    │   │   │       └── migrations
    │   │   │           ├── 20210831113004_torrust_users.sql
    │   │   │           ├── ...
    │   │   │           └── 20220308170028_torrust_categories.sql
    │   │   └── uploads
    │   │       ├── 1.torrent
    │   │       └── 2.torrent
    │   ├── mod.rs
    │   ├── output
    │   │   ├── destiny.db
    │   │   └── source.db
```

da2ce7 commented on migrations/mysql/20221109092556_torrust_user_date_registered_allow_null.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE torrust_users CHANGE date_registered date_registered DATETIME NOT NULL
```
This table alteration sets `NOT NULL`, shouldn't it be `DEFAULT NULL`?
Strange that a test didn't break with this...

josecelano replied:
Hi @da2ce7, good catch! I've changed only two things in the application behaviour:

1. The login works with both "argon2" and "pbkdf2". I tested it manually.
2. The "date_registered" is now optional (It can be NULL). I only tested the MySQL migrations by running manually "cargo test", but I did not test to create a new user without that date in MySQL because there is no migration for MySQL. There was no MySQL database in version v1.0.0. We could keep that NOT NULL option for MySQL, but I have to fix it to have the same schema as in SQLite.

josecelano replied:
I've added two new subtasks to the main PR description to test those changes.

da2ce7 commented:
@josecelano Amazing Work! I think that it will take a little time for me to review.

WarmBeer commented:
> The BitTorrent Protocol Specification v2 - BEP52

BitTorrent V2 torrents are just too different from V1 and barely used, so I haven't implemented support yet. All torrent files on  Index V1 should be BT V1.

`On the other hand, do you know why the md5sum field is always empty in TorrentInfo?:`

`md5sum` is a redundant field I believe.

da2ce7 commented:
> In conclusion, I do not think it makes sense to test it because after running the upgrade there is no way to create a user without a registration date. And the upgrader is executed just once.
> 
> On the other hand, we could test that the application does not allow empty registration date for new users. That can not happen now because the registration date is hardcoded in the SQL INSERT.

This test could be natural when we support migration from SQLite3 to MySQL. In the test data we could have users with a NULL registration date.

ldpr commented:
@josecelano I just wanted to drop a quick message to say thanks for this PR, we've north of 11 thousand torrents ATM (almost 19k total made!) so we badly needed this. We're extremely grateful and can't wait to test it out :)

ldpr commented:
I had some time to test this out and I've some feedback so figured it'd be good to drop it here cos it might affect other users too. Also, my apologies in advance, I have a fever and it's a little difficult to think clearly - let me know if anything is confusing!

When setting up Torrust the first time, we noticed that if we inserted a category capitalised (ie: ZIP), then it'd be saved as Zip.  We thought we made a mistake so we remade them and when we realised it was a torrust thing, we just accepted it. However, this resulted in our categories looking like this - `category_id` for 1-9 didn't exist and our actual categories were 10-15.

This caused the migration to error right at the start. I managed to work around this by manually editing the `category_id` in `torrust_categories` and `torrust_torrents` (moving categories 10-15 down to 1-6) _however_ I also ran into issues with the pragma foreign key as you noted at the beginning.

I tested with the migration afterwards and it worked great, everything appeared to be fine. _However_, only the first torrent in the database would show on pages like `recent`. Additionally, if we searched for a torrent we could find it, but it wouldn't display any information about the torrent. I'm pretty sure this is entirely my fault due to making the changes in the database to get past the category error noted above.

If there was a failsafe for non-existant categories it would likely fix it outright I think. 
Unfortunately I didn't save the output, if you think that'd be helpful let me know and I can start it over!

josecelano commented:
> I had some time to test this out and I've some feedback so figured it'd be good to drop it here cos it might affect other users too. Also apologies in advance, I have a fever and it's a little difficult to think clearly - let me know if anything is confusing!
> 
> When setting up Torrust the first time, we noticed that if we created a category capitalised (ie: ZIP), then it'd be saved as Zip. We thought we made a mistake so we remade them and when we realised it was a torrust thing, we just accepted it. However, this resulted in our categories looking like this - `category_id` for 1-9 didn't exist and our actual categories were 10-15.
> 
> This caused the migration to error right at the start. I managed to work around this by manually editing the `category_id` in `torrust_categories` and `torrust_torrents` (moving categories 10-15 down to 1-6) _however_ I also ran into issues with the pragma foreign key as you noted at the beginning.
> 
> I tested with the migration afterwards and it worked great, everything appeared to be fine. _However_, only the first torrent in the database would show on pages like `recent`. Additionally, if we searched for a torrent we could find it, but it wouldn't display any information about the torrent (in short, the torrent page was broken, no content displayed). I'm pretty sure this is entirely my fault due to making the changes in the database to get past the category error noted above.
> 
> If there was a failsafe for non-existant categories it would likely fix it outright I think. Unfortunately I didn't save the output, if you think that'd be helpful let me know and I can start it over!

hi @ldpr, thank you very much for your comment! I'm going to review it carefully but I think it was my mistake. For categories, I had two options:

1. Keep the id.
2. Reset the id sequence.

I chose the second option but I forgot to change the category id for the torrent. I'm using the name as the primary key because you cannot have two categories with the same name.

I'm going to fix it. I only need to get the new category id before inserting the new torrent.

By the way, @WarmBeer @da2ce7, in this case, I decided to reset the ids but for the other ids (for example, "users") I do not change them. Maybe I should be more consistent and change them always or never, what do you think? Maybe I should keep the old ID since users could have been using it for other tasks. @ldpr, which behaviour would you prefer? I suppose, in most cases, it does not matter.

da2ce7 commented:
@josecelano  It would be nice to keep the ID's unchanged if possible. However what is more important is that we are consistent.  Either we keep all the ID's, or None.  We shouldn't give an expectation that ID's are kept, but randomly are not.

ldpr commented:
For what it's worth @josecelano I'm absolutely fine with either option. Whichever solution looks best to you folks, works best for us too :)

josecelano commented:
> For what it's worth @josecelano I'm absolutely fine with either option. Whichever solution looks best to you folks, works best for us too :)

hi @ldpr, I have changed the script to keep category IDs in this [commit](https://github.com/torrust/torrust-index-backend/pull/77/commits/02d1ce335b72212b0b44a53833f3eb70f8567b8e).

Besides, I've added a test like your case using new categories, not the ones that are included by DB migrations.

ldpr commented:
Brilliant, everything looks great except for one minor thing. The migration worked great, no errors, users can signin/signup afterwards, previous torrents and new torrents work fine. Additionally, torrentpages now correctly show information.

Curiously, I'm still only seeing one torrent show on recent/popular and I'm not sure why. If i search a term that should bring up several results, only the first result from the search shows. If i make the search more precise, it finds the correct version. To be safe, I've tested with the `v2.0.0` frontend release also but no difference. 

Is it possible it's a frontend bug that hasn't been reported? @WarmBeer any ideas? There's no errors showing in browser console, nor any errors from the backend itself. 🤔

josecelano commented:
> Brilliant, everything looks great except for one minor thing. The migration worked great, no errors, users can signin/signup afterwards, previous torrents and new torrents work fine. Additionally, torrentpages now correctly show information.
> 
> Curiously, I'm still only seeing one torrent show on recent/popular and I'm not sure why. If i search a term that should bring up several results, only the first result from the search shows. If i make the search more precise, it finds the correct version. _However_, in all cases (using /recent/ or opening categories), the only torrent returned is the first torrent for each category (by `torrent_id`). To be safe, I've tested with the `v2.0.0` frontend release also but no difference.
> 
> Is it possible it's a frontend bug that hasn't been reported? @WarmBeer any ideas? There's no errors showing in browser console, nor any errors from the backend itself. thinking

hi @ldpr, I have the same problem. I have not had time to confirm it and find out why, but I'm assuming it's a bug in the latest version. I have also been experiencing some problems with the login form, but that could be a problem with the autofill feature in my browser.

josecelano commented:
@ldpr @WarmBeer I've opened a [new issue](https://github.com/torrust/torrust-index-backend/issues/84) describing how to reproduce the problem.

@ldpr, if you do not find any other problem regarding the migration script (not related to the data migration), I will merge this PR. I will do it as soon as I can discard that the problem was caused by something missing in the migration. Thank you very much for your help! It's been very valuable.

josecelano commented:
Hi @WarmBeer @da2ce7 @ldpr ,

- I have rebased the branch.
- I've updated the PR description with the new tasks "Subtasks from issue 84".

Since there is a bug in the current version that affects this migration script, I'm going to:

- Fix issue #84 
- Make some changes to the script.

We need to add data to the table `torrust_torrent_tracker_stats`. We have two options:

- We can create empty records for each torrent with 0 seeders and leechers.
- We can make a request to the tracker to pre-fill the number of seeders and leechers.

I think the second option is the expected behaviour after migration. That could make the script a little bit more complicated. If we have too many torrents, we need to make a lot of requests to the tracker, and some of them could timeout, etcetera. But I think that's something we already do every hour.

Alternatively, we could leave the table empty and write a NOTICE in the docs. You can change the configuration, set a polling interval to 1 second and import info for all torrents. And then re-set to the original 1-hour value.

I could also create a new binary to manually run that "sync" task so that you can run it manually after migrating the DB.

But I think the best option is to run the "sync task" automatically within the "upgrader".

Please let me know if you have any suggestions or alternatives.

da2ce7 commented on src/upgrades/from_v1_0_0_to_v2_0_0/databases/mod.rs:
```diff
@@ -0,0 +1,30 @@
+    Arc::new(SqliteDatabaseV2_0_0::new(&dest_database_connect_url).await)
+}
+
+pub async fn migrate_destiny_database(dest_database: Arc<SqliteDatabaseV2_0_0>) {
```
nit: maybe `destination` instead of `destiny`?

josecelano replied:
Yes, 😄 . That's the problem with using abbreviations like "dest"

da2ce7 replied:
maybe `target` then?

josecelano replied:
Yes, I also prefer "target". I was considering it-. I do not remember why I used "destiny". I will change it to "target".

josecelano replied:
Done

da2ce7 reviewed (commented):
Overall the changes look very good.

One more feature I suggest:  When a user runs this v2 on-top-of a v1 db.  It should detect this and point them to the v1 to v2 migration document.

josecelano commented:
> Overall the changes look very good.
> 
> One more feature I suggest: When a user runs this v2 on-top-of a v1 db. It should detect this and point them to the v1 to v2 migration document.

That would be great, but we have not yet discussed the best option to detect the software and DB version from the running software.

For the code, I suppose we have to use something like [this](https://docs.rs/built/latest/built/).

For the DB schema, the version could be the latest executed migration in the `_sql_migrations` table. In fact, we can use the `version` column.

I would suggest merging this PR as soon as I finish the pending subtasks and opening a new issue to detect DB and code version from the running software. Once that's done, we can modify this script to include that new feature. What do you think @da2ce7?

da2ce7 commented:
> I would suggest merging this PR as soon as I finish the pending subtasks and opening a new issue to detect DB and code version from the running software. Once that's done, we can modify this script to include that new feature. What do you think @da2ce7?

Okay. Lets do the DB detection in another issue.

josecelano commented:
ACK 5a7d87517cbaa8867a0a7f35a0a8a288205c361a

josecelano commented:
Hi @WarmBeer @da2ce7 @ldpr, I've changed my mind. Instead of importing stats automatically from the tracker, I've created a new console command to do it manually and optionally.

```s
cargo run --bin import_tracker_statistics
```

You can run that command after running the upgrader if you want to pre-fill the stats cache table (`torrust_torrent_tracker_stats`).

The reason is that after merging https://github.com/torrust/torrust-index-backend/pull/87 the app works fine again, showing all torrents on the list even if they do not have stats yet.

The app imports stats every hour for all torrents (settings option) when you upload a new torrent or when you load the torrent detail page. You can decide if you want to pre-populate the table or let the app fill it up.

da2ce7 reviewed (approved):
ACK 5a7d87517cbaa8867a0a7f35a0a8a288205c361a

josecelano commented on src/console/commands/import_tracker_statistics.rs:
```diff
@@ -0,0 +1,86 @@
+    eprintln!(
+        "{} - imports torrents statistics from linked tracker.
+
+        cargo run --bin upgrade SOURCE_DB_FILE DESTINY_DB_FILE TORRENT_UPLOAD_DIR
```
Fixed [here](https://github.com/torrust/torrust-index-backend/commit/83bafb4d1056411b0337b56a9449970d534c3178).
